aptos-move/aptos-vm/src/sharded_block_executor/sharded_aggregator_service.rs

```diff
@@ -0,0 +1,251 @@
+            (self.delta, rhs.delta)
+        } else {
+            (rhs.delta, self.delta)
+        };
+
+        if pos >= neg {
+            return Self {
+                delta: pos - neg,
+                is_positive: true,
+            };
+        }
+        Self {
+            delta: neg - pos,
+            is_positive: false,
+        }
+    }
+}
+
+#[test]
+fn test_delta_u128() {
+    assert_eq!(DeltaU128::default().delta, 0);
+    assert!(DeltaU128::default().is_positive);
+
+    {
+        // get positive delta
+        let delta_pos = DeltaU128::get_delta(100, 50);
+        assert_eq!(delta_pos.delta, 50);
+        assert!(delta_pos.is_positive);
+
+        // get negative delta
+        let delta_neg = DeltaU128::get_delta(50, 100);
+        assert_eq!(delta_neg.delta, 50);
+        assert!(!delta_neg.is_positive);
+    }
+
+    {
+        // test add_delta
+        let delta1 = DeltaU128 {
+            delta: 100,
+            is_positive: true,
+        };
+        assert_eq!(delta1.add_delta(50), 150);
+
+        let delta2 = DeltaU128 {
+            delta: 50,
+            is_positive: false,
+        };
+        assert_eq!(delta2.add_delta(50), 0);
+    }
+
+    {
+        // test all cases for ops::Add
+        let delta1 = DeltaU128 {
+            delta: 100,
+            is_positive: true,
+        };
+        let delta2 = DeltaU128 {
+            delta: 50,
+            is_positive: false,
+        };
+        let delta3 = DeltaU128 {
+            delta: 100,
+            is_positive: true,
+        };
+        // checks for [pos > neg]; [pos, pos]
+        let mut delta_sum = delta1 + delta2 + delta3;
+        assert_eq!(delta_sum.delta, 150);
+        assert!(delta_sum.is_positive);
+
+        let delta4 = DeltaU128 {
+            delta: 500,
+            is_positive: false,
+        };
+        let delta5 = DeltaU128 {
+            delta: 200,
+            is_positive: false,
+        };
+        // checks for [neg > pos]; [neg, neg]
+        delta_sum = delta_sum + delta4 + delta5;
+        assert_eq!(delta_sum.delta, 550);
+        assert!(!delta_sum.is_positive);
+    }
+}
+
+pub fn aggregate_and_update_total_supply<S: StateView>(
+    sharded_output: &mut Vec<Vec<Vec<TransactionOutput>>>,
+    global_output: &mut [TransactionOutput],
+    state_view: &S,
+    executor_thread_pool: Arc<rayon::ThreadPool>,
+) {
+    let num_shards = sharded_output.len();
+    let num_rounds = sharded_output[0].len();
+
+    // The first element is 0, which is the delta for shard 0 in round 0. +1 element will contain
+    // the delta for the global shard
+    let mut aggr_total_supply_delta = vec![DeltaU128::default(); num_shards * num_rounds + 1];
+
+    // No need to parallelize this as the runtime is O(num_shards)
+    // TODO: Get this from the individual shards while getting 'sharded_output'
+    sharded_output
+        .iter()
+        .enumerate()
+        .for_each(|(shard_id, shard_output)| {
```

Comment:
why not iterate all the rounds in the result txn version order and gather all the diffs in one go?

(better if we create a helper function `rounds_iter_in_version_order(&mut Vec<Vec<Vec<TransactionOutput>>>) -> Iterator<Item = &Vec<>>` )

Reply:
We are iterating such that we minimize going across shards. So once we go to a shard we iterate over all the rounds over there. 

If we iterate for each round first, and then every shard in each round, we could do the read and update of total supply in one pass over the Vec<Vec<Vec<TransactionOutput>>>, but it will require switching across shards for every round, which is costly since we eventually want shards across different nodes

Reply:
I don't quite buy this argument but my brain capacity (barely) works with two loops as well :trollface:

Reply:
As discussed, changed it to iterate round-wise for better readability